File: Misc/Find All Duplicates in an Array.java
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        
        ArrayList<Integer> result = new ArrayList<>();

        for (int i = 0; i < nums.length; i ++) {

            // use the SIGN of the current number's index (minus one, but 
            // you get the idea) to mark whether or not you've encountered
            // that number before

            // abs() in case it's negative
            int current_num = Math.abs(nums[i]);

            if (nums[current_num - 1] < 0) {
                // encountered this number before and marked the spot as negative
                // (prior encounter would have used the same formula)
                result.add(current_num);
            } else {
                nums[current_num - 1] *= -1;
            }
        }

        return result;

    }
}

File: Old Java Solutions/Misc/Find All Duplicates in an Array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:

        result = []

        for i, num in enumerate(nums):

            curr_num = abs(num)
            if nums[curr_num - 1] < 0:
                result.append(curr_num)
            else:
                nums[curr_num - 1] *= -1

        return result
